spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/DirectMessageListenerContainer.java

```diff
@@ -91,79 +98,243 @@ public void setTaskExecutor(TaskExecutor taskExecutor) {
 	 * @param consumersPerQueue the consumers per queue.
 	 */
 	public void setConsumersPerQueue(int consumersPerQueue) {
+		if (isRunning()) {
+			adjustConsumers(consumersPerQueue);
+		}
 		this.consumersPerQueue = consumersPerQueue;
 	}
 
 	/**
-	 * Set to false to requeue failed deliveries.
-	 * @param requeueRejected the requeue rejected.
+	 * Set to true for an exclusive consumer - if true, the
+	 * {@link #setConsumersPerQueue(int) consumers per queue} must be 1.
+	 * @param exclusive true for an exclusive consumer.
 	 */
-	public void setDefaultRequeueRejected(boolean requeueRejected) {
-		this.defaultRequeueRejected = requeueRejected;
+	@Override
+	public final void setExclusive(boolean exclusive) {
+		Assert.isTrue(!exclusive || (this.consumersPerQueue == 1),
+				"When the consumer is exclusive, the consumers per queue must be 1");
+		super.setExclusive(exclusive);
 	}
 
 	@Override
+	public void setQueueNames(String... queueName) {
+		Assert.state(!isRunning(), "Cannot set queue names while running, use add/remove");
+		super.setQueueNames(queueName);
+	}
+
+	@Override
+	public void setQueues(Queue... queues) {
+		Assert.state(!isRunning(), "Cannot set queue names while running, use add/remove");
+		super.setQueues(queues);
+	}
+
+	@Override
+	public void addQueueNames(String... queueNames) {
+		addQueues(Arrays.asList(queueNames));
+		super.addQueueNames(queueNames);
+	}
+
+	@Override
+	public void addQueues(Queue... queues) {
+		addQueues(Arrays.asList(queues)
+				.stream()
+				.map(Queue::getName)
+				.collect(Collectors.toList()));
+		super.addQueues(queues);
+	}
+
+	private void addQueues(Collection<String> queues) {
+		if (isRunning()) {
+			synchronized (this.consumersMonitor) {
+				Set<String> current = getQueueNamesAsSet();
+				for (String queue : queues) {
+					if (current.contains(queue)) {
+						throw new IllegalStateException("Queue " + queue + " is already configured for this container: "
+								+ this + ", no queues added");
+					}
+				}
+				List<String> added = new ArrayList<>();
+				for (String queue : queues) {
+					try {
+						consumeFromQueue(queue);
+						added.add(queue);
+					}
+					catch (IOException e) {
+						logger.error("Failed to add queue: " + queue + " undoing adds (if any)");
+						removeQueues(added);
+						throw new AmqpIOException("Failed to add " + queues, e);
+					}
+				}
+			}
+		}
+	}
+
+	@Override
+	public boolean removeQueueNames(String... queueNames) {
+		removeQueues(Arrays.asList(queueNames));
+		return super.removeQueueNames(queueNames);
+	}
+
+	@Override
+	public boolean removeQueues(Queue... queues) {
+		removeQueues(Arrays.asList(queues)
+				.stream()
+				.map(Queue::getName)
+				.collect(Collectors.toList()));
+		return super.removeQueues(queues);
+	}
+
+	private void removeQueues(Collection<String> queues) {
+		if (isRunning()) {
+			synchronized (this.consumersMonitor) {
+				for (String queue : queues) {
```

Comment:
It doesn't handle `null` returned from `remove()`, but I've done something similar.